compiler/rustc_resolve/src/late.rs

```diff
@@ -2401,32 +2401,70 @@ impl<'a: 'ast, 'b, 'ast, 'tcx> LateResolutionVisitor<'a, 'b, 'ast, 'tcx> {
                 });
             }
 
-            ItemKind::Static(box ast::StaticItem { ref ty, ref expr, .. })
-            | ItemKind::Const(box ast::ConstItem { ref ty, ref expr, .. }) => {
+            ItemKind::Static(box ast::StaticItem { ref ty, ref expr, .. }) => {
                 self.with_static_rib(|this| {
                     this.with_lifetime_rib(LifetimeRibKind::Elided(LifetimeRes::Static), |this| {
                         this.visit_ty(ty);
                     });
                     this.with_lifetime_rib(LifetimeRibKind::Elided(LifetimeRes::Infer), |this| {
                         if let Some(expr) = expr {
-                            let constant_item_kind = match item.kind {
-                                ItemKind::Const(..) => ConstantItemKind::Const,
-                                ItemKind::Static(..) => ConstantItemKind::Static,
-                                _ => unreachable!(),
-                            };
                             // We already forbid generic params because of the above item rib,
                             // so it doesn't matter whether this is a trivial constant.
                             this.with_constant_rib(
                                 IsRepeatExpr::No,
                                 ConstantHasGenerics::Yes,
-                                Some((item.ident, constant_item_kind)),
+                                Some((item.ident, ConstantItemKind::Static)),
                                 |this| this.visit_expr(expr),
                             );
                         }
                     });
                 });
             }
 
+            ItemKind::Const(box ast::ConstItem { ref generics, ref ty, ref expr, .. }) => {
+                // FIXME(generic_consts): Is the ordering of ribs correct? Right now, it's
+                // STATIC <- LIFETIME_ANON_REPORT <- GENERICS <- LIFETIME_ELIDED_STATIC {,<- CONSTANT}
+                self.with_static_rib(|this| {
+                    this.with_lifetime_rib(LifetimeRibKind::AnonymousReportError, |this| {
+                        this.with_generic_param_rib(
+                            &generics.params,
+                            RibKind::Item(HasGenericParams::Yes(generics.span)),
+                            LifetimeRibKind::Generics {
+                                binder: item.id,
+                                kind: LifetimeBinderKind::Item,
+                                span: generics.span,
+                            },
+                            |this| {
+                                this.with_lifetime_rib(
+                                    LifetimeRibKind::Elided(LifetimeRes::Static),
+                                    |this| {
+                                        this.visit_generics(generics);
```

Comment:
I'm not convinced that `visit_generics` should be inside this rib. But we don't have much precedent either, do we?

Reply:
I've now (locally) placed the `visit_generics` outside of `Elided(Static)`. In any case before & after that, rustc flagged elided lifetimes inside the generics of const items (where-clauses, defaults, const types). Not sure if I understand what difference it makes.